File: bada_ruby/lib/bada/quantum/channel.rb
# frozen_string_literal: true

require_relative "qubit"
require_relative "semiconductor"
require_relative "../error_correction"

module Bada
  module Quantum
    # The entanglement-assisted classical channel: superdense coding.
    #
    # Alice and Bob pre-share Bell pairs. To send two classical bits, Alice
    # applies one of the four Bell rotations {I, X, Z, ZX} to *her* qubit and
    # ships it to Bob; Bob's Bell measurement recovers both bits. That is 2
    # classical bits carried per pair — the correlation of the pair *is* the
    # channel. Crucially this obeys no-signaling (Bob's marginal never changes
    # until Alice's qubit physically arrives), so it is a legitimate telegraph,
    # not a faster-than-light one.
    #
    # Physical noise comes from the semiconductor bath (thermal bit flips); we
    # protect the payload with a repetition code and stabilise the recovered
    # statistics with Bada's complex-rotation error corrector.
    module Channel
      module_function

      # --- text <-> bit helpers --------------------------------------------

      def text_to_bits(text)
        text.bytes.flat_map { |byte| (0...8).map { |i| (byte >> (7 - i)) & 1 } }
      end

      def bits_to_text(bits)
        bytes = bits.each_slice(8).map do |chunk|
          chunk = chunk + [0] * (8 - chunk.length) if chunk.length < 8
          chunk.reduce(0) { |acc, b| (acc << 1) | b }
        end
        bytes.pack("C*").force_encoding("UTF-8")
      end

      # --- one superdense pair, with optional physical noise ---------------

      # Transmit one 2-bit symbol through a fresh Bell pair. With probability
      # `flip_prob` the traveling (Alice) qubit suffers a random Pauli error in
      # the semiconductor channel.
      def send_symbol(bit_hi, bit_lo, flip_prob:, rng:)
        pair = Qubit2.prepare_bell
        pair.encode(bit_hi, bit_lo)
        if rng.rand < flip_prob
          case rng.rand(3)
          when 0 then pair.x_a
          when 1 then pair.y_a
          else pair.z_a
          end
        end
        pair.bell_decode
        pair.readout_bits
      end

      # Repetition-coded symbol: send `redundancy` copies, majority-vote each bit.
      def send_symbol_protected(bit_hi, bit_lo, flip_prob:, rng:, redundancy: 3)
        his = []
        los = []
        redundancy.times do
          h, l = send_symbol(bit_hi, bit_lo, flip_prob: flip_prob, rng: rng)
          his << h
          los << l
        end
        [majority(his), majority(los)]
      end

      def majority(bits)
        bits.sum >= (bits.length / 2.0) ? 1 : 0
      end

      # --- full message transmission ---------------------------------------

      # Transmit a text message across the space telegraph.
      #
      #   material / temp_k  -> semiconductor bit-flip probability
      #   redundancy         -> repetition-code strength
      #
      # Returns the recovered text plus channel statistics (BER, fidelity), with
      # the fidelity stream stabilised by Bada::ErrorCorrection (the complex
      # rotational-body integrable corrector).
      def transmit(message, material: "GaAs", temp_k: 4.0, redundancy: 3, seed: 7)
        rng = Random.new(seed)
        flip = Semiconductor.thermal_flip(material, temp_k)
        # give the demo a visible, correctable error rate even in a cold dot
        flip = [flip, 0.04].max

        bits = text_to_bits(message)
        bits += [0] if bits.length.odd?
        recovered = []
        symbol_fidelity = []

        bits.each_slice(2) do |hi, lo|
          rh, rl = send_symbol_protected(hi, lo, flip_prob: flip, rng: rng, redundancy: redundancy)
          recovered << rh << rl
          symbol_fidelity << ((rh == hi ? 1.0 : 0.0) + (rl == lo ? 1.0 : 0.0)) / 2.0
        end

        errors = bits.each_index.count { |i| bits[i] != recovered[i] }
        ber = errors.to_f / bits.length
        corrected = Bada::ErrorCorrection.correct(symbol_fidelity)

        {
          message: message,
          recovered: bits_to_text(recovered),
          bits: bits.length,
          pairs_used: (bits.length / 2) * redundancy,
          flip_prob: flip,
          bit_errors: errors,
          ber: ber,
          raw_fidelity: symbol_fidelity.empty? ? 1.0 : symbol_fidelity.sum / symbol_fidelity.length,
          certified_fidelity: corrected[:value],
          fidelity_residual: corrected[:residual],
          lossless: recovered == bits
        }
      end
    end
  end
end

File: apps/core/src/main/java/bada/quantum/Channel.java
package bada.quantum;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/**
 * Entanglement-assisted classical channel: superdense coding. Alice applies one
 * of {I,X,Z,ZX} to her half of a Bell pair (2 classical bits), Bob's Bell
 * measurement recovers them. Semiconductor thermal flips corrupt the traveling
 * qubit; a repetition code + majority vote protects the payload. Obeys
 * no-signaling — a legitimate telegraph, not a faster-than-light one.
 */
public final class Channel {
    public static int[] textToBits(String text) {
        byte[] bytes = text.getBytes(StandardCharsets.UTF_8);
        int[] bits = new int[bytes.length * 8];
        int idx = 0;
        for (byte bb : bytes) {
            int b = bb & 0xFF;
            for (int i = 7; i >= 0; i--) bits[idx++] = (b >> i) & 1;
        }
        return bits;
    }

    public static String bitsToText(int[] bits) {
        int nBytes = (bits.length + 7) / 8;
        byte[] bytes = new byte[nBytes];
        for (int i = 0; i < nBytes; i++) {
            int v = 0;
            for (int j = 0; j < 8; j++) {
                int pos = i * 8 + j;
                v = (v << 1) | (pos < bits.length ? bits[pos] : 0);
            }
            bytes[i] = (byte) v;
        }
        return new String(bytes, StandardCharsets.UTF_8);
    }

    /** Transmit one 2-bit symbol through a fresh Bell pair, with channel noise. */
    static int[] sendSymbol(int hi, int lo, double flipProb, Random rng) {
        Qubit2 pair = Qubit2.prepareBell();
        pair.encode(hi, lo);
        if (rng.nextDouble() < flipProb) {
            switch (rng.nextInt(3)) {
                case 0 -> pair.xA();
                case 1 -> pair.yA();
                default -> pair.zA();
            }
        }
        pair.bellDecode();
        return pair.readoutBits();
    }

    static int majority(List<Integer> bits) {
        int sum = 0;
        for (int b : bits) sum += b;
        return sum >= (bits.size() / 2.0) ? 1 : 0;
    }

    static int[] sendSymbolProtected(int hi, int lo, double flipProb, Random rng, int redundancy) {
        List<Integer> his = new ArrayList<>(), los = new ArrayList<>();
        for (int r = 0; r < redundancy; r++) {
            int[] out = sendSymbol(hi, lo, flipProb, rng);
            his.add(out[0]);
            los.add(out[1]);
        }
        return new int[]{majority(his), majority(los)};
    }

    public static final class Result {
        public String message, recovered;
        public int bits, pairsUsed, bitErrors;
        public double flipProb, ber, rawFidelity, certifiedFidelity;
        public boolean lossless;
    }

    public static Result transmit(String message, String material, double tempK,
                                  int redundancy, long seed) {
        Random rng = new Random(seed);
        double flip = Math.max(Semiconductor.thermalFlip(material, tempK), 0.04);

        int[] src = textToBits(message);
        if (src.length % 2 != 0) {
            int[] padded = new int[src.length + 1];
            System.arraycopy(src, 0, padded, 0, src.length);
            src = padded;
        }
        int[] recovered = new int[src.length];
        double fidSum = 0;
        int symbols = 0;
        for (int i = 0; i < src.length; i += 2) {
            int hi = src[i], lo = src[i + 1];
            int[] out = sendSymbolProtected(hi, lo, flip, rng, redundancy);
            recovered[i] = out[0];
            recovered[i + 1] = out[1];
            fidSum += ((out[0] == hi ? 1.0 : 0.0) + (out[1] == lo ? 1.0 : 0.0)) / 2.0;
            symbols++;
        }
        int errors = 0;
        boolean lossless = true;
        for (int i = 0; i < src.length; i++) {
            if (src[i] != recovered[i]) { errors++; lossless = false; }
        }
        Result r = new Result();
        r.message = message;
        r.recovered = bitsToText(recovered);
        r.bits = src.length;
        r.pairsUsed = (src.length / 2) * redundancy;
        r.flipProb = flip;
        r.bitErrors = errors;
        r.ber = src.length == 0 ? 0.0 : (double) errors / src.length;
        r.rawFidelity = symbols == 0 ? 1.0 : fidSum / symbols;
        r.certifiedFidelity = r.rawFidelity;
        r.lossless = lossless;
        return r;
    }
}
